sycl/source/detail/device_impl.cpp

```diff
@@ -8,19 +8,97 @@
 
 #include <CL/sycl/detail/device_impl.hpp>
 #include <CL/sycl/device.hpp>
+#include <algorithm>
 
 namespace cl {
 namespace sycl {
 namespace detail {
 
+device_impl::device_impl() : MIsHostDevice(true) {}
+
+device_impl::device_impl(RT::PiDevice Device)
+    : MDevice(Device), MIsHostDevice(false) {
+  // TODO catch an exception and put it to list of asynchronous exceptions
+  PI_CALL(RT::piDeviceGetInfo(MDevice, PI_DEVICE_INFO_TYPE,
+                              sizeof(RT::PiDeviceType), &MType, 0));
+
+  RT::PiDevice parent;
+  // TODO catch an exception and put it to list of asynchronous exceptions
+  PI_CALL(RT::piDeviceGetInfo(MDevice, PI_DEVICE_INFO_PARENT,
+                              sizeof(RT::PiDevice), &parent, 0));
+
+  MIsRootDevice = (nullptr == parent);
+  if (!MIsRootDevice) {
+    // TODO catch an exception and put it to list of asynchronous exceptions
+    PI_CALL(RT::piDeviceRetain(MDevice));
+  }
+}
+
+device_impl::~device_impl() {
+  if (!MIsRootDevice && !MIsHostDevice) {
+    // TODO catch an exception and put it to list of asynchronous exceptions
+    CHECK_OCL_CODE_NO_EXC(RT::piDeviceRelease(MDevice));
+  }
+}
+
+bool device_impl::is_affinity_supported(
+    info::partition_affinity_domain AffinityDomain) const {
+  auto SupportedDomains = get_info<info::device::partition_affinity_domains>();
+  return std::find(SupportedDomains.begin(), SupportedDomains.end(),
+                   AffinityDomain) != SupportedDomains.end();
+}
+
+cl_device_id device_impl::get() const {
+  if (MIsHostDevice)
+    throw invalid_object_error("This instance of device is a host instance");
+
+  if (!MIsRootDevice) {
+    // TODO catch an exception and put it to list of asynchronous exceptions
+    PI_CALL(RT::piDeviceRetain(MDevice));
+  }
+  // TODO: check that device is an OpenCL interop one
+  return pi::cast<cl_device_id>(MDevice);
+}
+
+platform device_impl::get_platform() const {
+  if (MIsHostDevice)
+    return platform();
```

Comment:
Do we always create new platform for the host device? Looks like a bug as it makes the following assert fails:

sycl::device Dev(sycl::host_selector());
assert(Dev.get_platform() == Dev.get_platform());

PS. This comment doesn't ask changes.

Reply:
@romanovvlad for non-host devices new object is created either. See

https://github.com/intel/llvm/blob/7e8db8f56465a0f7837e36d5d5768548605c079b/sycl/include/CL/sycl/detail/common.hpp#L124-L130

While I agree this looks bad, I think it is a bit out of scope of this change list.